external-crates/move/crates/move-compiler/src/typing/syntax_methods.rs

```diff
@@ -0,0 +1,234 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module verifies the usage of the "syntax method" functions. These functions are declared
+//! as 'syntax' but have not been ensured to be type-compatible or otherwise adhere to our
+//! trait-like constraints around their definitions. We process them here, using typing machinery
+//! to ensure the are.
+
+use crate::{
+    diag,
+    expansion::ast::AbilitySet,
+    expansion::ast::ModuleIdent,
+    ice,
+    naming::ast::{self as N, IndexSyntaxMethods, SyntaxMethod},
+    typing::core::{self, Context},
+};
+use move_ir_types::location::*;
+
+//-------------------------------------------------------------------------------------------------
+// Validation
+//-------------------------------------------------------------------------------------------------
+
+pub fn validate_syntax_methods(
+    context: &mut Context,
+    _mident: &ModuleIdent,
+    module: &mut N::ModuleDefinition,
+) {
+    let methods = &mut module.syntax_methods;
+    for (_, entry) in methods.iter_mut() {
+        if let Some(index) = &mut entry.index {
+            let IndexSyntaxMethods { index, index_mut } = &mut **index;
+            if let (Some(index_defn), Some(index_mut_defn)) = (index.as_ref(), index_mut.as_ref()) {
+                if !validate_index_syntax_methods(context, index_defn, index_mut_defn) {
+                    // If we didn't validate they wre comptaible, we remove the mut one to avoid more
+                    // typing issues later.
+                    assert!(context.env.has_errors());
+                    *index_mut = None;
+                }
+            }
+        }
+    }
+}
+
+fn validate_index_syntax_methods(
+    context: &mut Context,
+    index: &SyntaxMethod,
+    index_mut: &SyntaxMethod,
+) -> bool {
+    let index_ann_loc = index.kind.loc;
+    let (index_module, index_fn) = &index.target_function;
+    let (index_mut_module, index_mut_fn) = &index_mut.target_function;
+
+    let index_finfo = context.function_info(index_module, index_fn).clone();
+    let mut_finfo = context
+        .function_info(index_mut_module, index_mut_fn)
+        .clone();
+
+    if index_finfo.signature.type_parameters.len() != mut_finfo.signature.type_parameters.len() {
+        let index_msg = format!(
+            "This index function expects {} type arguments",
+            index_finfo.signature.type_parameters.len()
+        );
+        let index_mut_msg = format!(
+            "This mutable index function expects {} type arguments",
+            mut_finfo.signature.type_parameters.len()
+        );
+        let mut diag = diag!(
+            TypeSafety::IncompatibleSyntaxMethods,
+            (index.loc, index_msg),
+            (index_mut.loc, index_mut_msg),
+        );
+        diag.add_note(
+            "Index operations on the same type must take the name number of type arguments",
```

Comment:
Should we also check the same constraints?

Reply:
I intentionally left the constraints open. It's actually why I perform the skolomeziation instead of simply unifying one set of type parameters with the other's instantiation. The constraints will all be sorted out when we try to use these functions, and it seemed kinder to the programmer to allow polymorphic type constraints to vary based on usage. If you'd like me to address that, please let me know.

Reply:
Fair enough, yeah I suppose we don't need to enforce them to be similar or not because the constraints can be applied after we determine the usage. Seems fine to me